wallet/src/Cardano/Wallet/WalletLayer/Kernel.hs

```diff
@@ -153,6 +170,30 @@ bracketPassiveWallet pm mode logFunction keystore node fInjects f = do
         invokeIO :: forall m'. MonadIO m' => Actions.WalletAction Blund -> m' ()
         invokeIO = liftIO . STM.atomically . invoke
 
+-- takes a WalletDecrCredentials and transaction, and returns the Coin output, if its ours
+maybeReadcoin :: (HdRootId, WalletDecrCredentials) -> (TxIn, TxOutAux) -> Maybe Coin
+maybeReadcoin wkey (_, txout) = case isOurs (txOutAddress . toaOut $ txout) [wkey] of
+  (Just _, _)  -> Just $ (txOutValue . toaOut) txout
+  (Nothing, _)-> Nothing
+
+-- take a EncryptedSecretKey and return the sum of all utxo in the state, and the walletid
+xqueryWalletBalance :: Kernel.PassiveWallet -> EncryptedSecretKey -> IO WalletBalance
+xqueryWalletBalance w esk = do
+  let
+    nm = makeNetworkMagic (w ^. walletProtocolMagic)
+  let
+    rootid = eskToHdRootId nm esk
+    wdc = eskToWalletDecrCredentials nm esk
+  let
+    withNode :: (HasCompileInfo, HasUpdateConfiguration) => Lock (WithNodeState IO) -> WithNodeState IO [Coin]
+    withNode _lock = filterUtxo (maybeReadcoin (rootid, wdc))
+  my_coins <- withNodeState (w ^. walletNode) withNode
+  let
+    balance :: Coin
+    balance = foldl' unsafeAddCoin (mkCoin 0) my_coins
+    walletid = addrToBase58 $ view fromDb (getHdRootId rootid)
```

Comment:
I'd recommend to use [Conv.toRootId](https://github.com/input-output-hk/cardano-sl/blob/35e8f14b88733a39e4abc7e06d55562ef10986ee/wallet/src/Cardano/Wallet/WalletLayer/Kernel/Conv.hs#L138-L139) despite its poorly chosen name :/